### PR TITLE
Feature: Add Note Velocity Editor to Automation View

### PR DESCRIPTION
**Pull request by seangoodvibes**
## **Feature: Automation View Note Velocity Editor**

New Note Velocity Editing View that has been added as part of the existing Automation View implementation so that you can edit the velocities and other parameters of notes in a single note row. 

Screenshot 2024-06-05 at 11 34 47 AM

### **Description:**

- imagine if the whole grid was just one note row, or one kit row
- each column of colours represents a note
- the vertical colours represents that notes velocity which you can edit by interacting with the grid (like you would with a parameter in automation view)

### **Outstanding Items:**

<details><summary>Future PR Work</summary>

- [ ] Code Review + Refactoring
- [ ] Fix note editor state so that it is clip specific and not view specific (e.g. entering note editor in one clip should not cause another clip with automation view open to enter note editor).
- [ ] Fix selected note row so that it is clip specific and not view specific
- [ ] Bugfix - Might still be an issue with 7seg display but can't reproduce. 
- [ ] Fix any bugs / tweaks needed for velocity ramps
- [ ] Add ability to shift velocity ramp up down
- [ ] Add velocity ramp selection to pad selection mode
- [ ] Add ability to switch from left to right pad while velocity ramp is selected
- [ ] Activate gold knobs for adjusting velocity ramp when it is selected?
- [ ] Refactor audition pad action in automation view to call code in instrument clip view as there is duplication
- [ ] LastAuditionedYDisplay can become out of sync with selected drum in a kit
</details>
<details><summary>Completed</summary>

- [x] Fix note row selection blinking when re-entering automation view after exiting
- [x] Fix grey out of velocity editor when entering menu
- [x] Fix Shift + <> to shorten / extend the note row's length and not the clip's length while in note editor
- [x] Change note deletion to happen when you press pad that corresponds to squares average velocity
- [x] Change bottom pad press to set velocity to 1 instead of deleting note
- [x] Fix <> + back to clear only the selected note row's notes and not the entire clip
- [x] Enter velocity view with audition pad + velocity shortcut from instrument clip view
- [x] Fix note probability editing - you can now press and hold on the top velocity pad in a column to edit the note's probability
- [x] Fix note repeat editing - you can now press and hold on the top velocity pad in a column to edit the note's repeats
- [x] Enable cross screen velocity note editing
- [x] Fix 7SEG display rendering
- [x] Bugfix - Switching from pad selection mode in note editor to another param keeps you in pad selection mode, but it seems like left pad selection is stuck. Fix it by exiting pad selection mode b4 switching to other params with shortcut.
- [x] Add quantize / humanize combo (audition pad + press and turn tempo)
- [x] Add velocity ramps - Change velocity of all notes between two presses
- [x] Velocity ramp - Fixed bug where OLED display would not show L and R position of Velocity ramp even while still multi pad press is active
- [x] Fix notes recorded in with audition pads not rendering until you turn off recording
- [x] Fix 7Seg display rendering for velocity ramps
- [x] Update row probability so that you can edit row probability by just turning select while you are outside pad selection mode and aren't currently holding a note press
- [x] Re-added note row transpose shortcut (Press + turn vertical encoder)
- [x] Add shortcuts for rotating note row press vertical + press and turn <> and press audition + Press and turn <>
- [x] Add horizontal scrolling and zooming into pad selection mode
- [x] Add waveform shortcut blinking to indicate you're in pad selection mode
- [x] Fix pad selection mode can't be enabled while in menu view
- [x] Fix euclidean - it removes notes entered by previous euclidean actions (e.g. enter euclidean notes on one note row, then switch to another note row and add notes with euclidean and then remove them - it will removes notes from current note row and previous note row - I think it's because it's doing an "undo" action
- [x] Bugfix - Sometimes note row selected and velocity shortcut doesn't blink (noticed it when jumping from automation editor to velocity editor) - Also noticed it exiting out of velocity editor into instrument clip view and back into velocity view
</details>

### **Quick Overview:**
- Enter the Velocity Editor View by:
   - While in the Automation Instrument Clip View, use Shift + Velocity or press Velocity pad on Automation Overview to enter the Velocity Note Editor
   OR
   - While in the Instrument Clip View, hold Audition pad + Press Velocity shortcut pad
- Select note row:
  - Press audition pad or Vertical scroll to select different note rows
  - The audition pad will blink to indicate the currently selected note row
  - The display will indicate the note code / drum name of the row you have selected
- Add Note / Adjust Velocity Using Grid:
  - Press grid to set/adjust velocities of existing notes and add note (if no note exists)
- Set a Velocity Ramp:
  - Press and hold two grid note pads to set a velocity ramp between the two notes. 
  - Note: for this to work, you must first press a pad that has a note in it and then press a second pad that has a note in it. if the second pad has no note already in it, it will set a tail instead for the first note pad pressed (or if it's a drum sound, like a kick, it will just add another note in the position of the second pad press).
- Remove Note:
  - Short press top velocity pad in a note column to remove note
- Select Note and Adjust Note Parameters:
  - Long press top velocity pad in a note column to see note's velocity and perform note related adjustment actions (e.g. adjust velocity, probability). Note press related actions include:
     - **Adjust Velocity:** Turn horizontal encoder <>
     - **Adjust Probability:** Turn Select Encoder
     - **Adjust Note Repeat:** Press and Turn Vertical Encoder
     - **Adjust Automation:** Turn gold knobs
- Adjust Note Row Transpose:
  - Press and turn vertical encoder
  - Note: does not work if you are holding a note (transpose gets replaced with editing note repeats)
- Adjust Note Row Probability:
  - Hold Audition Pad + Turn Select Encoder to change the selected Note Row's probability
     - Note: does not work if you are currently holding a note
- Add/Remove Notes with Euclidean:
   - Hold Audition Pad + Press and turn vertical encoder
   - Note: The Euclidean action is replaced by note repeat editing if you have selected a note (either by holding a note outside pad selection mode or if you are in pad selection mode)
- Change clip length:
   - Hold Shift + turn horizontal encoder <>
- Change selected note row length by:
   - Holding Audition pad + Turning horizontal encoder <>
   - Note: does not work if you have selected a note
- Clear notes in selected note row by: 
   - Holding horizontal encoder <> + pressing Back
   - Note: does not work if you have selected a note
- Rotate Note Row by:
   - Holding vertical encoder + turning horizontal encoder <>
   - Holding audition pad + pressing and turning horizontal encoder <>
   - Note: does not work if you have selected a note
- Quantize / Humanize notes by:
   - Enabling Quantize community feature
   - Hold audition pad + press and turn tempo encoder to set quantize / humanize %
- Cross Screen Editing:
    - Press cross screen while in the Velocity Note Editor to edit notes across multiple screens. 
    - Note: you can only toggle cross screen editing if the clip has multiple screens OR the current note row selected has multiple screens. 

### ** Future PR Functionality: **

The following functionality was removed from this PR as it still needs work.

- Pad Selection Mode:
  - Press Shift + Waveform to enter Pad Selection Mode
  - Pad Selection Mode allows you to select note and perform all the same actions as if you were holding that note's pad. This includes:
    - Turn <> to adjust the selected note's velocity (in pad selection mode you can also hold shift to adjust the note's velocity more quickly)
    - Press Shift + Turn <> to adjust selected notes velocity faster
    - Press + Turn vertical encoder to adjust the selected note's repeats
    - Turn select encoder to adjust the selected note's probability
    - Note: to use the following actions in Pad Selection Mode, you need to first de-select a note by pressing that note again:
      - Horizontal scrolling (Turn <>)
      - Zooming (Press + Turn <>)

### **Some Demo Videos:**

Euclidean Fun:

https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/7bae10b3-d577-47a2-be66-22e4ea43a1bf

Playing with Probability:

https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/651c0939-32f4-4f10-932d-d4dc5750dc23

Euclidean and Probability Together:

https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/08ae6680-646e-4dcc-b6a4-552790fdcfab

Enter Velocity Ramps for Existing Notes:

https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/8d171c36-b52c-403b-8413-7655b44912fa

### CODE REVIEW

**Comment by m-m-adams** [comment]
Featurewise my last nit is that holding a note and turning <> encoder should fine tune velocity (same as it does outside velocity automation view)

File: src/deluge/gui/views/instrument_clip_view.cpp [hunk 1]
```diff
@@ -3992,7 +4039,8 @@ int32_t InstrumentClipView::setupForEnteringScaleMode(int32_t newRootNote, int32
 
 	// Need to figure out the scale first...
 	getCurrentInstrumentClip()->inScaleMode = true;
-	currentSong->setRootNote(newRootNote, getCurrentInstrumentClip()); // Computation to find out what notes in scale
+	currentSong->setRootNote(newRootNote,
+	                         getCurrentInstrumentClip()); // Computation to find out what notes in scale
```
Comment (m-m-adams) [inline src/deluge/gui/views/instrument_clip_view.cpp:4043]: weird wrapping

File: src/deluge/gui/views/instrument_clip_view.cpp [hunk 2]
```diff
@@ -5184,7 +5245,17 @@ void InstrumentClipView::graphicsRoutine() {
 	uint8_t nonMutedColour = clip->getCurrentlyRecordingLinearly() ? 2 : 0;
 	for (int32_t yDisplay = 0; yDisplay < kDisplayHeight; yDisplay++) {
 		int32_t noteRowIndex;
-		NoteRow* noteRow = clip->getNoteRowOnScreen(yDisplay, currentSong, &noteRowIndex);
+		NoteRow* noteRow = nullptr;
+		// if we're in the automation view note editor, then we're only looking at one note row at a time
+		// so we want to render a single note row's playhead across all note rows
+		if (getRootUI() == &automationView && automationView.inNoteEditor()) {
+			noteRow = clip->getNoteRowOnScreen(lastAuditionedYDisplay, currentSong, &noteRowIndex);
```
Comment (m-m-adams) [inline src/deluge/gui/views/instrument_clip_view.cpp:5252]: this seems inefficient - probably put this outside the loop so we don't repeatedly grab the noterow on screen while updating

File: src/deluge/gui/views/instrument_clip_view.cpp [hunk 3]
```diff
@@ -5622,28 +5694,32 @@ void InstrumentClipView::rotateNoteRowHorizontally(ModelStackWithNoteRow* modelS
...
+		// or also shift Automation when default setting to only shift automation in Automation View is false
+		bool shiftAutomation = ((currentUI == &automationView && !automationView.inNoteEditor())
+		                        || (currentUI != &automationView && !FlashStorage::automationShift));
 
-		// Always clear Notes and MPE when you're not in Automation View
-		bool shiftSequenceAndMPE = (currentUI != &automationView);
+		// If you're in Automation View, only shift Notes and MPE if you're in the Note Editor
+		// Always shift Notes and MPE when you're not in Automation View
+		bool shiftSequenceAndMPE =
+		    ((currentUI != &automationView) || (currentUI == &automationView && automationView.inNoteEditor()));
```
Comment (m-m-adams) [inline src/deluge/gui/views/instrument_clip_view.cpp:5705]: there's enough differences here that I almost feel like note automation should be its own view?

Comment (m-m-adams) [reply src/deluge/gui/views/instrument_clip_view.cpp:5705]: Lots of checks for whether you're in note editor or not in this PR

File: src/deluge/model/note/note_row.cpp [hunk 4]
```diff
@@ -194,6 +194,167 @@ Error NoteRow::beenCloned(ModelStackWithNoteRow* modelStack, bool shouldFlattenR
...
 }
 
+// initialize row square info array before we iterate through the note row to store square info
+void NoteRow::initRowSquareInfo(SquareInfo rowSquareInfo[kDisplayWidth], bool anyNotes) {
+	for (int32_t x = 0; x < kDisplayWidth; x++) {
+		initSquareInfo(rowSquareInfo[x], anyNotes, x);
+	}
+}
+
+void NoteRow::initSquareInfo(SquareInfo& squareInfo, bool anyNotes, int32_t x) {
```
Comment (m-m-adams) [inline src/deluge/model/note/note_row.cpp:204]: this should be a constructor on SquareInfo that takes anyNotes and x as arguments

File: src/deluge/gui/views/instrument_clip_view.h [hunk 5]
```diff
@@ -226,20 +233,16 @@ class InstrumentClipView final : public ClipView, public InstrumentClipMinder {
...
 	void deleteCopiedNoteRows();
 
 	void createDrumForAuditionedNoteRow(DrumType drumType);
 	void nudgeNotes(int32_t offset);
-	void editNoteRepeat(int32_t offset);
 
 	bool isRowAuditionedByInstrument(int32_t yDisplay);
 
-	void editNumEuclideanEvents(ModelStackWithNoteRow* modelStack, int32_t offset, int32_t yDisplay);
 	void rotateNoteRowHorizontally(ModelStackWithNoteRow* modelStack, int32_t offset, int32_t yDisplay,
```
Comment (m-m-adams) [inline src/deluge/gui/views/instrument_clip_view.h:246]: this should probably be doable in velocity view too

**Review by m-m-adams — COMMENTED** [review]
Automation view should be split into two, just make a note automation view that inherits from automation view and overrides the functions that are checking inNoteEditor() right now. It'll reduce the spaghetti factor by a ton and also run more efficiently

File: src/deluge/gui/views/instrument_clip_view.cpp (hunk 3 above)
Comment (seangoodvibes) [reply src/deluge/gui/views/instrument_clip_view.cpp:5792]: Will split out the views in the next PR.

**Review by m-m-adams — APPROVED** [review]

**Pull request merged**